osquery/tables/applications/darwin/browser_plugins.cpp

```diff
@@ -227,26 +281,204 @@ QueryData genSafariExtensions(QueryContext& context) {
+
+  // Iterate over the list of user extensions and check if there is a match with
+  // extension identifier
+  getPtreeFromPlist(app_extension_plist, app_extension_ptree);
+  for (const auto& entry : app_extension_ptree) {
+    if (!boost::algorithm::contains(entry.first, ext_data.identifier)) {
+      return true;
+    }
+  }
+
+  getPtreeFromPlist(web_extension_plist, web_extension_ptree);
+  for (const auto& entry : web_extension_ptree) {
+    if (!boost::algorithm::contains(entry.first, ext_data.identifier)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+inline void genSafariSandboxedExtensions(const QueryContext& context,
+                                         QueryData& results) {
+  // Getting user data
+  auto users = usersFromContext(context);
+  for (const auto& row : users) {
+    auto uid = row.at("uid");
+    auto user_dir = row.at("directory");
+
+    // Sanity check
+    if (uid.empty() || user_dir.empty()) {
+      continue;
+    }
+
+    // Skipping non-user queries
+    if (!boost::algorithm::starts_with(user_dir, "/Users")) {
+      continue;
+    }
+
+    // We need to get the sandboxed extension data first
+
+    // Check that an extensions directory exists.
+    if (!pathExists(kSafariAppExtensionsPath).ok()) {
+      return;
+    }
+
+    // Getting app directories
+    std::vector<std::string> app_directories;
+    if (!listDirectoriesInDirectory(
+            kSafariAppExtensionsPath, app_directories, false)) {
+      return;
+    }
+
+    // Traverse app directories to obtain app extension data if present
+    SandboxedExtensionsData sandboxed_ext_data;
+    for (const auto& app_directory : app_directories) {
+      std::vector<std::string> app_internal_dirs;
+      if (!listDirectoriesInDirectory(app_directory, app_internal_dirs, true)) {
+        continue;
+      }
+
+      // Get global sandboxed extensions data
+      for (const auto& app_int_dir : app_internal_dirs) {
+        if (boost::algorithm::ends_with(app_int_dir,
+                                        kSafariSandboxedExtensionsPattern)) {
+          getSandboxedExtensionData(
+              app_directory, app_int_dir + "Info.plist", sandboxed_ext_data);
+        }
+      }
+    }
+
+    // Return if no sandboxed extensions were found - No extensions installed
+    if (sandboxed_ext_data.empty()) {
+      return;
+    }
+
+    // We now have the sandboxed extension metainformation, we now need to check
+    // if extensions were installed by current user
+    auto user_app_extensions_plist =
+        fs::path(user_dir) / kAppExtensionsPlistPath;
+
+    auto user_web_extensions_plist =
+        fs::path(user_dir) / kWebExtensionsPlistPath;
+
+    // Check if app extension was installed by a given user
+    for (const auto& extension_data : sandboxed_ext_data) {
+      if (isUserExtension(user_app_extensions_plist,
+                          user_web_extensions_plist,
+                          extension_data)) {
```

Comment:
This is accidentally quadratic. If I read the for loop right:
For each user; for each app; for each plist file

We should not crawl /applications for each user. And we should not parse the plist files anew each time.

Reply:
This should be fixed now. The crawling of /Applications is performed once on new logic.